pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/BaseAndOrBooleanFilterOptimizer.java

```diff
@@ -0,0 +1,141 @@
+      case NOT:
+        // Recursively traverse the expression tree to find an operator node that can be rewritten.
+        operands.forEach(operand -> optimize(operand, schema));
+
+        // We have rewritten the child operands, so rewrite the parent if needed.
+        return optimizeCurrent(filterExpression);
+      default:
+        return optimizeChild(filterExpression, schema);
+    }
+  }
+
+  abstract boolean canBeOptimized(Expression filterExpression, @Nullable Schema schema);
+
+  /**
+   * Optimize any cases that are not AND/OR/NOT. This should be done by converting any cases
+   * that are always true to TRUE or always false to FALSE.
+   */
+  abstract Expression optimizeChild(Expression filterExpression, @Nullable Schema schema);
+
+  /**
+   * If any of the operands of AND function is "false", then the AND function itself is false and can be replaced with
+   * "false" literal. Otherwise, remove all the "true" operands of the AND function. Similarly, if any of the operands
+   * of OR function is "true", then the OR function itself is true and can be replaced with "true" literal. Otherwise,
+   * remove all the "false" operands of the OR function.
+   */
+  protected Expression optimizeCurrent(Expression expression) {
+    Function function = expression.getFunctionCall();
+    String operator = function.getOperator();
+    List<Expression> operands = function.getOperands();
+    if (operator.equals(FilterKind.AND.name())) {
+      // If any of the literal operands are always false, then replace AND function with FALSE.
+      for (Expression operand : operands) {
+        if (isAlwaysFalse(operand)) {
+          return FALSE;
+        }
+      }
+
+      // Remove all Literal operands that are always true.
+      operands.removeIf(this::isAlwaysTrue);
+      if (operands.isEmpty()) {
+        return TRUE;
+      }
+    } else if (operator.equals(FilterKind.OR.name())) {
+      // If any of the literal operands are always true, then replace OR function with TRUE
+      for (Expression operand : operands) {
+        if (isAlwaysTrue(operand)) {
+          return TRUE;
+        }
+      }
+
+      // Remove all Literal operands that are always false.
+      operands.removeIf(this::isAlwaysFalse);
+      if (operands.isEmpty()) {
+        return FALSE;
+      }
+    } else if (operator.equals(FilterKind.NOT.name())) {
+      assert operands.size() == 1;
+      Expression operand = operands.get(0);
+      if (isAlwaysTrue(operand)) {
+        return FALSE;
+      }
+      if (isAlwaysFalse(operand)) {
+        return TRUE;
+      }
+    }
+    return expression;
+  }
+
+  private boolean isAlwaysFalse(Expression operand) {
+    return operand.equals(FALSE);
+  }
+
+  private boolean isAlwaysTrue(Expression operand) {
+    return operand.equals(TRUE);
+  }
+
+  /** Change the expression value to boolean literal with given value. */
+  protected static void setExpressionToBoolean(Expression expression, boolean value) {
```

Comment:
Not introduced in this PR, but let's remove this method since we should avoid mutating an expression. We can use the constant `TRUE` and `FALSE` instead

Reply:
done. There's still some more mutation in NumericalFilterOptimizer, but this gets rid of a big part